Add mixed precision interfaces

huppd opened the pull request:
This add generic interfaces that allows to pass single or double precission to ecrad_gas.

reuterbal reviewed (commented):
Apologies for letting this hang for so long.

In principle this looks good - however, since everything is essentially kept in `jprb` precision and relies on implicit casting of `pressure_hl` to `jprb` from whatever precision it was provided in, I'm wondering if it wouldn't be better to avoid the duplication of the routines. Instead, we could create the precision specific interfaces and then call a `jprb`-version with an explicit cast. I'll try to outline below:


```f90
type cloud_type

module test_kind_mod
  use iso_fortran_env, only: real64, real32
  implicit none

  integer, parameter :: jprb = real32 ! real64

  type cloud_type
    real(jprb), allocatable :: pressure(:,:)
  contains
    procedure :: eta_jprb
    procedure :: eta_real32
    procedure :: eta_real64
    generic :: eta => eta_real32, eta_real64
  end type cloud_type

contains

  subroutine eta_jprb(this, pressure_hl, nlon, nlev)
    class(cloud_type) :: this
    real(jprb), intent(in) :: pressure_hl(:,:)
    integer, intent(in) :: nlon, nlev
    integer :: jl, jk
    do jl=1,nlon
      do jk=1,nlev
        this%pressure(jl,jk) = pressure_hl(jl,jk+1) - pressure_hl(jl, jk)
      end do
    end do
  end subroutine eta_jprb

  subroutine eta_real32(this, pressure_hl, nlon, nlev)
    class(cloud_type) :: this
    real(real32), intent(in) :: pressure_hl(:,:)
    integer, intent(in) :: nlon, nlev
    call this%eta_jprb(real(pressure_hl, kind=jprb), nlon, nlev)
  end subroutine eta_real32

  subroutine eta_real64(this, pressure_hl, nlon, nlev)
    class(cloud_type) :: this
    real(real64), intent(in) :: pressure_hl(:,:)
    integer, intent(in) :: nlon, nlev
    call this%eta_jprb(real(pressure_hl, kind=jprb), nlon, nlev)
  end subroutine eta_real64
end module test_kind_mod

program test_kind
  use test_kind_mod, only: cloud_type
  use iso_fortran_env, only: real32, real64

  type(cloud_type) :: cloud
  real(real32), allocatable :: pressure_hl(:,:)
  integer :: klon, klev, jk, jl

  klon = 4
  klev = 5

  allocate(pressure_hl(klon, klev+1))
  allocate(cloud%pressure(klon, klev))

  do jl=1,klon
    do jk=1,klev+1
      pressure_hl(jl, jk) = real(jl + jk, kind=real32)
    end do
  end do

  call cloud%eta(pressure_hl,klon,klev)

  do jl=1,klon
    do jk=1,klev
      if (cloud%pressure(jl,jk) .ne. 1._real32) then
        write(0,*) 'Unexpected pressure(jl,jk):',cloud%pressure(jl,jk),' at ',jl,',',jk
      end if
    end do
  end do
end program

```

huppd commented:
> Apologies for letting this hang for so long.
> 
> In principle this looks good - however, since everything is essentially kept in `jprb` precision and relies on implicit casting of `pressure_hl` to `jprb` from whatever precision it was provided in, I'm wondering if it wouldn't be better to avoid the duplication of the routines. Instead, we could create the precision specific interfaces and then call a `jprb`-version with an explicit cast. I'll try to outline below:
> 
> ```fortran-modern
> type cloud_type
> 
> module test_kind_mod
>   use iso_fortran_env, only: real64, real32
>   implicit none
> 
>   integer, parameter :: jprb = real32 ! real64
> 
>   type cloud_type
>     real(jprb), allocatable :: pressure(:,:)
>   contains
>     procedure :: eta_jprb
>     procedure :: eta_real32
>     procedure :: eta_real64
>     generic :: eta => eta_real32, eta_real64
>   end type cloud_type
> 
> contains
> 
>   subroutine eta_jprb(this, pressure_hl, nlon, nlev)
>     class(cloud_type) :: this
>     real(jprb), intent(in) :: pressure_hl(:,:)
>     integer, intent(in) :: nlon, nlev
>     integer :: jl, jk
>     do jl=1,nlon
>       do jk=1,nlev
>         this%pressure(jl,jk) = pressure_hl(jl,jk+1) - pressure_hl(jl, jk)
>       end do
>     end do
>   end subroutine eta_jprb
> 
>   subroutine eta_real32(this, pressure_hl, nlon, nlev)
>     class(cloud_type) :: this
>     real(real32), intent(in) :: pressure_hl(:,:)
>     integer, intent(in) :: nlon, nlev
>     call this%eta_jprb(real(pressure_hl, kind=jprb), nlon, nlev)
>   end subroutine eta_real32
> 
>   subroutine eta_real64(this, pressure_hl, nlon, nlev)
>     class(cloud_type) :: this
>     real(real64), intent(in) :: pressure_hl(:,:)
>     integer, intent(in) :: nlon, nlev
>     call this%eta_jprb(real(pressure_hl, kind=jprb), nlon, nlev)
>   end subroutine eta_real64
> end module test_kind_mod
> 
> program test_kind
>   use test_kind_mod, only: cloud_type
>   use iso_fortran_env, only: real32, real64
> 
>   type(cloud_type) :: cloud
>   real(real32), allocatable :: pressure_hl(:,:)
>   integer :: klon, klev, jk, jl
> 
>   klon = 4
>   klev = 5
> 
>   allocate(pressure_hl(klon, klev+1))
>   allocate(cloud%pressure(klon, klev))
> 
>   do jl=1,klon
>     do jk=1,klev+1
>       pressure_hl(jl, jk) = real(jl + jk, kind=real32)
>     end do
>   end do
> 
>   call cloud%eta(pressure_hl,klon,klev)
> 
>   do jl=1,klon
>     do jk=1,klev
>       if (cloud%pressure(jl,jk) .ne. 1._real32) then
>         write(0,*) 'Unexpected pressure(jl,jk):',cloud%pressure(jl,jk),' at ',jl,',',jk
>       end if
>     end do
>   end do
> end program
> ```

No problem, as discussed I removed one of the interface changed and extracted the checks from the other interface. I am open for better names. Let me know if I can do anything else.

reuterbal reviewed (approved):
Many thanks for the changes, looks great now!
